File: src/pi.py
from decimal import Decimal, setcontext, Context
from time import time


def main() -> None:
    ITERATIONS = 200
    ITERATIONS_FOR_AVG = 30
    PRECISION = 50
    setcontext(Context(prec=PRECISION))
    print("Hello World, From Python!")
    output = None
    time_sum_seconds = 0
    for i in range(ITERATIONS_FOR_AVG):
        start_time_seconds = time()
        output = pi_from_nilakantha_series(ITERATIONS)
        end_time_seconds = time()
        time_sum_seconds += end_time_seconds - start_time_seconds
    average_time_seconds = time_sum_seconds / ITERATIONS_FOR_AVG
    average_time_ms = int(average_time_seconds * 1000)
    average_time_ns = int(average_time_seconds * 1000000000)
    print(
        f"Value of pi calcualted with the Nilakantha Series: {output}.\nAverage Time of execution (in nanoseconds): {average_time_ns}.\nAverage Time of execution (in milliseconds): {average_time_ms}.\nNumber of Iterations: {ITERATIONS}."
    )


def pi_from_nilakantha_series(iterations: int) -> Decimal:
    pi = Decimal(3.0)
    for i in range(1, iterations + 1):
        pi += (1 if i % 2 == 1 else -1) * (
            Decimal(4) / ((2 * i) * (2 * i + 1) * (2 * i + 2))
        )
    return pi


main()

File: src/pi.rs
use std::time::{Duration,Instant};
fn main() -> () {
    let iterations = 50;
    let iterations_for_avg = 100;
    println!("Hello World, From Rust!");
    let mut total_duration = Duration::new(0,0);
    for _iteration in 0..20{
        let start = Instant::now();
        calculate_pi_newton_euler_convergence(iterations);
        total_duration += start.elapsed();
    } 
    let time_ns = total_duration.as_nanos()/iterations_for_avg;
    println!("Value of pi calculated with the Newton/Euler convergence: {} \nAverage Time Taken (nanoseconds): {} \nNumber of Iterations: {}",calculate_pi_newton_euler_convergence(50), time_ns, iterations);
}

fn calculate_pi_newton_euler_convergence(iterations: u32) -> f64 {
    let mut sum = 0.0_f64;
    for n in 0..iterations{
        // println!("{}",((2_f64).powf(n.into()) * factorial(n) * factorial(n))/factorial(2*n + 1));
        sum += ((2_f64).powf(n.into()) * factorial(n) * factorial(n))/factorial(2*n + 1);
    } // maybe simplify this later by combining factorial series to allow for more iterations
    // can't simplify easily (would require doing more floating point operations so I think it would actually increase execution time)
    return sum*2_f64;

}

fn factorial(i: u32) -> f64 {
    let mut result = 1_f64;
    for n in 1..=i{
        result *= f64::from(n);
    }
    return result;
}
